staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go

```diff
@@ -142,6 +143,11 @@ func WithImpersonation(handler http.Handler, requestContextMapper request.Reques
 			}
 		}
 
+		if !groupsSpecified && username != user.Anonymous {
```

Comment:
Now I'm puzzled. Why we need to do this, if we have [this snippet](https://github.com/kubernetes/kubernetes/blob/022919d1a4051fe39861fc5491967aea394b4787/staging/src/k8s.io/apiserver/pkg/audit/request.go#L76-L92) in place and audit happens before impersonation?

Reply:
We have 4 [stages](https://github.com/kubernetes/kubernetes/blob/bc352342693e337b1ae5094af70265419a1247c5/staging/src/k8s.io/apiserver/pkg/apis/audit/v1beta1/types.go#L56).
This change will effect the other three stages rather than "StageRequestReceived".

We do this because [the snippet](https://github.com/kubernetes/kubernetes/blob/022919d1a4051fe39861fc5491967aea394b4787/staging/src/k8s.io/apiserver/pkg/audit/request.go#L76-L92) may not complete.

Reply:
Reference:
https://github.com/kubernetes/kubernetes/pull/48184/files/3989d45adfccd57ffe1fdab15d3e04ab0262e152#diff-4bcea545afa7a35798d0ff9db0ad0e1aR121

Reply:
As far as I see the groups and UID might be different than in the old code. Can we drop the old snippet and use this instead? Having two code paths for the same things is confusing.

Reply:
👍 to what Stefan said, replace the old code with the new one, entirely. I did not understand you correctly, initially.

Reply:
Done.